File: apps/web/src/app/core/dimensions/band-config-store.ts
import { Signal, signal } from '@angular/core';

export interface BandConfigState<T> {
  version: number;
  byBand: Record<string, T>; // keyed by DimensionBand.id
}

// Generic per-face config keyed by dimension-band id. Subclasses provide the
// storage key, version, and default fields per band. Persistence + immutable
// updates are shared. Subclasses MUST call `this.init()` from their constructor
// (after `super()`), so injected dependencies are available before loading.
export abstract class BandConfigStore<T> {
  protected abstract storageKey(): string;
  protected abstract version(): number;
  protected abstract buildDefaults(): Record<string, T>;

  private readonly _state = signal<BandConfigState<T>>({ version: 0, byBand: {} });
  readonly state: Signal<BandConfigState<T>> = this._state.asReadonly();

  protected init(): void {
    this._state.set(this.load());
  }

  // Current fields for a band id. Defensive: a band id that exists in the
  // registry but is missing from byBand (the two persist separately) falls back
  // to a freshly built default, never undefined (it feeds varsFor).
  config(bandId: string): T {
    const byBand = this._state().byBand;
    return byBand[bandId] ?? this.buildDefaults()[bandId] ?? Object.values(byBand)[0];
  }

  protected patch(bandId: string, fn: (t: T) => T): void {
    const byBand = { ...this._state().byBand, [bandId]: fn(this.config(bandId)) };
    this.commit({ ...this._state(), byBand });
  }

  protected patchAll(fn: (t: T) => T): void {
    const byBand: Record<string, T> = {};
    for (const [id, fields] of Object.entries(this._state().byBand)) byBand[id] = fn(fields);
    this.commit({ ...this._state(), byBand });
  }

  private load(): BandConfigState<T> {
    try {
      const raw = localStorage.getItem(this.storageKey());
      if (raw) {
        const parsed = JSON.parse(raw) as BandConfigState<T>;
        if (parsed?.byBand) return this.migrate(parsed);
      }
    } catch {
      // fall through to seed
    }
    const seeded: BandConfigState<T> = { version: this.version(), byBand: this.buildDefaults() };
    this.persist(seeded);
    return seeded;
  }

  // Combine one band's default with its persisted value. Default: take the
  // persisted band wholesale (band-id-level merge — historical behavior).
  // Override to do field-level migration (fill new fields, translate shapes).
  protected mergeBand(defaults: T, persisted: T): T {
    return persisted ?? defaults;
  }

  // Additive: every default band id is present; a persisted band is combined
  // with its default via mergeBand (so new fields can be filled). Never wipes.
  protected migrate(state: BandConfigState<T>): BandConfigState<T> {
    const defaults = this.buildDefaults();
    const byBand: Record<string, T> = {};
    for (const id of Object.keys(defaults)) {
      byBand[id] = id in state.byBand ? this.mergeBand(defaults[id], state.byBand[id]) : defaults[id];
    }
    // Carry any persisted band ids not in defaults (forward-compat), untouched.
    for (const id of Object.keys(state.byBand)) {
      if (!(id in byBand)) byBand[id] = state.byBand[id];
    }
    const migrated: BandConfigState<T> = { version: this.version(), byBand };
    if (
      state.version !== this.version() ||
      Object.keys(byBand).length !== Object.keys(state.byBand).length
    ) {
      this.persist(migrated);
    }
    return migrated;
  }

  private commit(state: BandConfigState<T>): void {
    this._state.set(state);
    this.persist(state);
  }

  private persist(state: BandConfigState<T>): void {
    try {
      localStorage.setItem(this.storageKey(), JSON.stringify(state));
    } catch {
      // quota/unavailable — in-memory only
    }
  }
}

File: packages/AllyClockCore/Sources/AllyClockCore/Dimensions/BandConfigStore.swift
import Foundation

public struct BandConfigState<T: Codable>: Codable {
    public var version: Int
    public var byBand: [String: T]
    public init(version: Int, byBand: [String: T]) { self.version = version; self.byBand = byBand }
}

/// Generic per-face config keyed by `DimensionBand.id`. Subclasses supply the
/// storage key, version, and per-band defaults; persistence and immutable
/// updates are shared. Port of `band-config-store.ts`. Persists to UserDefaults
/// as JSON. Overridable hooks are pure (no subclass stored state), so calling
/// them from `init` is safe.
open class BandConfigStore<T: Codable> {
    private let defaults: UserDefaults
    public private(set) var state: BandConfigState<T>

    open var storageKey: String { fatalError("subclass must override storageKey") }
    open var version: Int { fatalError("subclass must override version") }
    open func buildDefaults() -> [String: T] { fatalError("subclass must override buildDefaults") }

    /// Combine one band's default with its persisted value. Default: take the
    /// persisted value wholesale. Override for field-level migration.
    open func mergeBand(default d: T, persisted p: T) -> T { p }

    public init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
        self.state = BandConfigState(version: 0, byBand: [:])
        self.state = load()
    }

    /// Current fields for a band id, never nil: falls back to a fresh default,
    /// then to any present band.
    public func config(_ bandId: String) -> T {
        state.byBand[bandId] ?? buildDefaults()[bandId] ?? state.byBand.values.first!
    }

    public func patch(_ bandId: String, _ transform: (T) -> T) {
        var byBand = state.byBand
        byBand[bandId] = transform(config(bandId))
        commit(BandConfigState(version: state.version, byBand: byBand))
    }

    public func patchAll(_ transform: (T) -> T) {
        var byBand: [String: T] = [:]
        for (id, fields) in state.byBand { byBand[id] = transform(fields) }
        commit(BandConfigState(version: state.version, byBand: byBand))
    }

    // MARK: - Persistence

    private func load() -> BandConfigState<T> {
        if let data = defaults.data(forKey: storageKey),
           let parsed = try? JSONDecoder().decode(BandConfigState<T>.self, from: data),
           !parsed.byBand.isEmpty {
            return migrate(parsed)
        }
        let seeded = BandConfigState(version: version, byBand: buildDefaults())
        persist(seeded)
        return seeded
    }

    /// Additive: every default band id is present; a persisted band is combined
    /// with its default via `mergeBand`. Persisted-only ids are carried untouched.
    private func migrate(_ persisted: BandConfigState<T>) -> BandConfigState<T> {
        let defs = buildDefaults()
        var byBand: [String: T] = [:]
        for id in defs.keys {
            if let p = persisted.byBand[id] { byBand[id] = mergeBand(default: defs[id]!, persisted: p) }
            else { byBand[id] = defs[id]! }
        }
        for (id, p) in persisted.byBand where byBand[id] == nil { byBand[id] = p }
        let migrated = BandConfigState(version: version, byBand: byBand)
        if persisted.version != version || byBand.count != persisted.byBand.count { persist(migrated) }
        return migrated
    }

    private func commit(_ next: BandConfigState<T>) { state = next; persist(next) }

    private func persist(_ state: BandConfigState<T>) {
        if let data = try? JSONEncoder().encode(state) { defaults.set(data, forKey: storageKey) }
    }
}
